File: src/pocketcloud/cloud/traffic/TrafficMonitor.php
<?php

namespace pocketcloud\cloud\traffic;

use Closure;

abstract class TrafficMonitor {

    public const string REGULAR_MODE_IN = "in";
    public const string REGULAR_MODE_OUT = "out";
    public const string SUFFIX_AVG = "_avg";

    protected bool $active = true;
    protected int $timestamp;
    protected ?int $monitoringDuration = null;

    /** @var array<array<Closure>> */
    protected array $handlers = [];

    protected int $totalBytesIn = 0;
    protected int $totalBytesOut = 0;
    protected array $byteHistoryIn = [];
    protected array $byteHistoryOut = [];

    protected ?Closure $stopMonitoringHandler = null;

    public function __construct(protected readonly string $monitorType) {
        $this->timestamp = time();
    }

    protected function addHandler(string $mode, Closure $handler): void {
        if (!isset($this->handlers[$mode])) $this->handlers[$mode] = [];
        $this->handlers[$mode][] = $handler;
    }

    /**
     * @param Closure $handler function (string $buffer, int $bytesIn, Address $source): void;
     * @return self
     */
    public function monitorIn(Closure $handler): self {
        $this->addHandler(self::REGULAR_MODE_IN, $handler);
        return $this;
    }

    /**
     * @param Closure $handler function (string $buffer, int $bytesOut, Address $destination): void;
     * @return self
     */
    public function monitorOut(Closure $handler): self {
        $this->addHandler(self::REGULAR_MODE_OUT, $handler);
        return $this;
    }

    /** @internal */
    public function pushBytes(string $mode, int $bytes): void {
        if (!$this->active) return;
        $now = microtime(true);
        switch (strtolower($mode)) {
            case self::REGULAR_MODE_IN: {
                $this->totalBytesIn += $bytes;
                $this->byteHistoryIn[] = [$now, $bytes];
                break;
            }
            case self::REGULAR_MODE_OUT: {
                $this->totalBytesOut += $bytes;
                $this->byteHistoryOut[] = [$now, $bytes];
                break;
            }
        }
    }

    /** @internal */
    public function cleanupHistory(): void {
        $now = microtime(true);
        $threshold = $now - 1;

        foreach ($this->byteHistoryIn as $i => $data) {
            if ($data[0] < $threshold) unset($this->byteHistoryIn[$i]);
        }

        foreach ($this->byteHistoryOut as $i => $data) {
            if ($data[0] < $threshold) unset($this->byteHistoryOut[$i]);
        }
    }

    public function callHandlers(string $mode, mixed ...$args): void {
        if (!$this->active) return;
        if (isset($this->handlers[$mode])) {
            foreach ($this->handlers[$mode] as $handler) {
                $handler->call($this, ...$args);
            }
        }
    }

    public function registerStopMonitoringHandler(?Closure $handler): void {
        $this->stopMonitoringHandler = $handler;
    }

    public function onStopMonitoring(mixed ...$args): bool {
        return false;
    }

    final public function stopMonitoring(mixed ...$args): void {
        if (!$this->active) return;
        $this->active = false;
        $this->handlers = [];
        $this->monitoringDuration = time() - $this->timestamp;
        TrafficMonitorManager::getInstance()->removeTrafficMonitor($this);
        if (!$this->onStopMonitoring(...$args) && $this->stopMonitoringHandler !== null) ($this->stopMonitoringHandler)(...$args);
    }

    public function isActive(): bool {
        return $this->active;
    }

    public function getTimestamp(): int {
        return $this->timestamp;
    }

    public function getMonitoringDuration(): int {
        if ($this->monitoringDuration === null) return time() - $this->timestamp;
        return $this->monitoringDuration;
    }

    public function getHandlers(): array {
        return $this->handlers;
    }

    public function getTotalBytesIn(): int {
        return $this->totalBytesIn;
    }

    public function getAverageBytesIn(): int {
        return array_sum(array_map(fn(array $data) => $data[1], $this->byteHistoryIn));
    }

    public function getTotalBytesOut(): int {
        return $this->totalBytesOut;
    }

    public function getTotalBytes(): int {
        return $this->totalBytesOut + $this->totalBytesIn;
    }

    public function getAverageBytesOut(): int {
        return array_sum(array_map(fn(array $data) => $data[1], $this->byteHistoryOut));
    }

    public function getAverageTotalBytes(): int {
        return $this->getAverageBytesOut() + $this->getAverageBytesIn();
    }

    public function getMonitorType(): string {
        return $this->monitorType;
    }
}

File: src/main/java/de/pocketcloud/cloud/traffic/TrafficMonitor.java
package de.pocketcloud.cloud.traffic;

import de.pocketcloud.cloud.util.TimeUtils;
import de.pocketcloud.cloud.util.TriConsumer;
import lombok.Getter;

import java.net.SocketAddress;
import java.util.Deque;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;

public abstract class TrafficMonitor {

    public static final String REGULAR_MODE_IN = "in";
    public static final String REGULAR_MODE_OUT = "out";
    public static final String SUFFIX_AVG = "_avg";

    @Getter
    protected final AtomicBoolean active = new AtomicBoolean(true);
    @Getter
    protected final long timestamp;
    protected volatile Long monitoringDuration = null;
    @Getter
    protected final Map<String, List<TriConsumer<SocketAddress, Object, Long>>> handlers = new ConcurrentHashMap<>();
    @Getter
    protected final AtomicLong totalBytesIn = new AtomicLong();
    @Getter
    protected final AtomicLong totalBytesOut = new AtomicLong();
    protected final Deque<TrafficMonitorManager.TrafficData> byteHistoryIn = new ConcurrentLinkedDeque<>();
    protected final Deque<TrafficMonitorManager.TrafficData> byteHistoryOut = new ConcurrentLinkedDeque<>();
    protected volatile Consumer<Object[]> stopMonitoringHandler = null;
    @Getter
    private final String monitorType;

    public TrafficMonitor(String monitorType) {
        this.monitorType = monitorType;
        this.timestamp = TimeUtils.currentSeconds();
    }

    public TrafficMonitor monitorIn(TriConsumer<SocketAddress, Object, Long> handler) {
        addHandler(REGULAR_MODE_IN, handler);
        return this;
    }

    public TrafficMonitor monitorOut(TriConsumer<SocketAddress, Object, Long> handler) {
        addHandler(REGULAR_MODE_OUT, handler);
        return this;
    }

    protected void addHandler(String mode, TriConsumer<SocketAddress, Object, Long> handler) {
        handlers.computeIfAbsent(mode, _ -> new CopyOnWriteArrayList<>())
                .add(handler);
    }

    public void pushBytes(String mode, long bytes) {
        if (!active.get()) return;
        double now = TimeUtils.currentTime();

        switch (mode.toLowerCase()) {
            case REGULAR_MODE_IN -> {
                totalBytesIn.addAndGet(bytes);
                byteHistoryIn.add(new TrafficMonitorManager.TrafficData(now, bytes));
            }
            case REGULAR_MODE_OUT -> {
                totalBytesOut.addAndGet(bytes);
                byteHistoryOut.add(new TrafficMonitorManager.TrafficData(now, bytes));
            }
        }
    }

    public void cleanupHistory() {
        double threshold = TimeUtils.currentTime() - 1.0;

        while (true) {
            TrafficMonitorManager.TrafficData first = byteHistoryIn.peekFirst();

            if (first == null || first.time() >= threshold) {
                break;
            }

            byteHistoryIn.pollFirst();
        }

        while (true) {
            TrafficMonitorManager.TrafficData first = byteHistoryOut.peekFirst();

            if (first == null || first.time() >= threshold) {
                break;
            }

            byteHistoryOut.pollFirst();
        }
    }

    public void callHandlers(String mode, SocketAddress address, Object buffer, Long bytes) {
        if (!active.get()) return;
        List<TriConsumer<SocketAddress, Object, Long>> modeHandlers = handlers.get(mode);

        if (modeHandlers == null) return;

        modeHandlers.forEach(m -> m.accept(address, buffer, bytes));
    }

    public void registerStopMonitoringHandler(Consumer<Object[]> handler) {
        this.stopMonitoringHandler = handler;
    }

    public final void stopMonitoring(Object... args) {
        if (!active.compareAndSet(true, false)) return;
        handlers.clear();
        monitoringDuration = TimeUtils.currentSeconds() - timestamp;

        TrafficMonitorManager.instance().removeTrafficMonitor(this);

        Consumer<Object[]> handler = stopMonitoringHandler;
        if (!onStopMonitoring(args) && handler != null) {
            handler.accept(args);
        }
    }

    public boolean onStopMonitoring(Object... args) {
        return false;
    }

    public long getMonitoringDuration() {
        Long duration = monitoringDuration;

        if (duration != null) {
            return duration;
        }

        return TimeUtils.currentSeconds() - timestamp;
    }

    public long getTotalBytes() {
        return totalBytesOut.get() + totalBytesIn.get();
    }

    public long getAverageTotalBytes() {
        return getAverageBytesOut() + getAverageBytesIn();
    }

    public long getAverageBytesOut() {
        return byteHistoryOut.stream()
                .mapToLong(TrafficMonitorManager.TrafficData::bytes)
                .sum();
    }

    public long getAverageBytesIn() {
        return byteHistoryIn.stream()
                .mapToLong(TrafficMonitorManager.TrafficData::bytes)
                .sum();
    }
}